src/pudl/output/eia860.py

```diff
@@ -464,3 +485,616 @@ def ownership_eia860(pudl_engine, start_date=None, end_date=None):
+        .assign(
+            unit_id_pudl=lambda x: x.max_unit_id_pudl + 1,
+            bga_source=lambda x: "agg_complex_cc",
+        )
+        .drop("max_unit_id_pudl", axis="columns")
+    )
+
+    out_df = gens_df.loc[~row_mask].append(
+        gens_df.loc[row_mask]
+        .drop(["unit_id_pudl", "bga_source"], axis="columns")
+        .merge(
+            unit_ids,
+            on="plant_id_eia",
+            how="left",
+            validate="many_to_one",
+        )
+    )
+    return out_df
+
+
+def flag_orphan_combined_cycle_gens(gens_df, orphan_code):
+    """
+    Flag CA/CT generators without corresponding CT/CAs as orphans.
+
+    Identify CA and CT generators that don't have Unit IDs and also have no
+    corresponding generator of the other type within the plant, that could be
+    used to create a whole combined-cycle unit, and flag them as orphans. In
+    theory this shouldn't happen, but the data isn't perfect. In some cases it
+    appears that the steam parts may have been incorrectly given prime mover
+    codes of ST or the combustion turbines may have been labeled GT.
+
+    Args:
+        gens_df (pandas.DataFrame): Must contain the columns plant_id_eia,
+            generator_id, and prime_mover_code. The column bga_code will
+            be created or assigned in matching rows.
+        orphan_code (str): Prime mover code indicating which type of generator
+            is being flagged as orphaned. Must be either "CA" or "CT".
+
+    Returns:
+        pandas.DataFrame
+
+    Raises:
+        ValueError: if orphan_code is not CA or CT.
+
+    """
+    if orphan_code == "CA":
+        missing_code = "CT"
+    elif orphan_code == "CT":
+        missing_code = "CA"
+    else:
+        raise ValueError(
+            f"orphan must be either 'CA' or 'CT', but we got {orphan_code}."
+        )
+
+    orphan_plant_ids = (
+        gens_df[gens_df.unit_id_pudl.isnull()]
+        .drop_duplicates(subset=["plant_id_eia", "generator_id", "prime_mover_code"])
+        .groupby(["plant_id_eia", "prime_mover_code"])
+        .size()
+        .unstack(fill_value=0)
+        .query(f"{orphan_code} > 0 & {missing_code} == 0")
+    ).index
+    row_mask = (
+        (gens_df.plant_id_eia.isin(orphan_plant_ids))
+        & (gens_df.unit_id_pudl.isnull())
+        & (gens_df.prime_mover_code == orphan_code)
+    )
+    out_df = gens_df.loc[~row_mask].append(
+        gens_df.loc[row_mask]
+        .assign(bga_source=f"orphan_{orphan_code.lower()}")
+    )
+    return out_df
+
+
+def assign_prime_fuel_unit_ids(gens_df, prime_mover_code, fuel_type_code_pudl):
+    """
+    Assign a PUDL Unit ID to all generators with a given prime mover and fuel.
+
+    This only assigns a PUDL Unit ID to generators that don't already have one,
+    and only to generators that have a consistent `fuel_type_code_pudl` across
+    all of the years of data in `gens_df`. This is a simplified fuel code that
```

Comment:
can you add the why in here for this one? I get the idea of what is happening, but why does the consistency of the fuel type matter?

Reply:
This really only applies to the Steam Turbines. That type of prime mover is just too generic -- basically everything ends up spinning a steam turbine -- and the fuel type is the only way we have to differentiate meaningfully between distinct "units" when there are several steam turbines that are part of a single plant.  Does that make sense?